qiskit/scheduler/methods/basic.py

```diff
@@ -140,31 +142,47 @@ def translate_gates_to_pulse_defs(circuit: QuantumCircuit,
     circ_pulse_defs = []
 
     cmd_def = schedule_config.cmd_def
-    measured_qubits = set()  # Collect qubits that would like to be measured
+    qubit_mem_slots = {}  # Map measured qubit index to classical bit index
 
     def get_measure_schedule() -> CircuitPulseDef:
         """Create a schedule to measure the qubits queued for measuring."""
         measures = set()
         all_qubits = set()
         sched = Schedule()
-        for q in measured_qubits:
-            measures.add(tuple(schedule_config.meas_map[q]))
+        for qubit in qubit_mem_slots:
+            measures.add(tuple(schedule_config.meas_map[qubit]))
         for qubits in measures:
             all_qubits.update(qubits)
-            # TODO (Issue #2704): Respect MemorySlots from the input circuit
-            sched |= cmd_def.get('measure', qubits)
-        measured_qubits.clear()
+            unused_mem_slots = set(qubits) - set(qubit_mem_slots.values())
+            default_sched = cmd_def.get('measure', qubits)
+            for time, inst in default_sched.instructions:
+                if isinstance(inst, AcquireInstruction):
+                    mem_slots = []
+                    for channel in inst.acquires:
+                        if channel.index in qubit_mem_slots.keys():
+                            mem_slots.append(MemorySlot(qubit_mem_slots[channel.index]))
+                        else:
+                            mem_slots.append(MemorySlot(unused_mem_slots.pop()))
+                    new_acquire = AcquireInstruction(command=inst.command,
+                                                     acquires=inst.acquires,
+                                                     mem_slots=mem_slots)
+                    sched |= new_acquire << time
+                # Measurement pulses should only be added if its qubit was measured by the user
+                elif inst.channels[0].index in qubit_mem_slots.keys():
+                    sched |= inst << time
```

Comment:
I will dislike this notation. Everytime I have to check in the code what it means. Maybe explicit `union` and `shift`? I understand that can be sugar syntax for the user, but there is no need to use it internally.

Reply:
Fair point. I would love to talk more about this with you... I've had several discussions about syntax improvements. You are not the only person who feels like this is not intuitive.